tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -840,10 +834,44 @@ private[optimizer] abstract class OptimizerCore(
       case tree: BinaryOp =>
         pretransformBinaryOp(tree)(cont)
 
+      case tree: JSBracketSelect =>
+        pretransformJSBracketSelect(tree)(cont)
+
       case tree: JSFunctionApply =>
         pretransformJSFunctionApply(tree, isStat = false,
             usePreTransform = true)(cont)
 
+      case JSArrayConstr(items) =>
+        if (items.exists(_.isInstanceOf[JSSpread])) {
+          /* TODO This means spread in array constr does not compose under
+           * this optimization. We could improve this with a
+           * pretransformExprsOrSpreads() or something like that.
+           */
+          cont(JSArrayConstr(transformExprsOrSpreads(items)).toPreTransform)
+        } else {
+          pretransformExprs(items) { titems =>
+            tryOrRollback { cancelFun =>
```

Comment:
Do we have any empirical data about how often this will happen in a codebase closer to JS? I'm a bit concerned that this causes too many rollbacks.